feat(auth): add username with password auth support

tysoekong opened the pull request.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/resty-redis-cluster?pullRequest=13) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Kong/resty-redis-cluster?pullRequest=13) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/resty-redis-cluster?pullRequest=13) it.</sub>

Tieske commented on kong-redis-cluster-1.4.0-0.rockspec:
```diff
@@ -1,8 +1,8 @@
 package = "kong-redis-cluster"
-version = "1.3.0-1"
```
please remove these changes. These belong in the release commit, not with code changes.

Tieske commented on kong-redis-cluster-1.4.0-0.rockspec:
```diff
@@ -1,8 +1,8 @@
+version = "1.4.0-0"
 source = {
     url = "git://github.com/Kong/resty-redis-cluster",
-    tag = "1.3.0-1"
```
The `-1` at the end should not be changed, it is the rockspec revision, it normally starts at 1. Mood anyway, as this has to be removed from this commit.

It is only presenyt in this tag here, because we had to release a rockspec change  in the past. Upon the next version bump the tag should be brought back to simply "x.y.z" without the revision.

Tieske reviewed (dismissed):
@outsinre besides the version bumps, this looks good to me. Would you mind having a look as well? (since James is no longer with Kong, I saw you as the only one having made code changes since then)

tysoekong commented:
Can we release this one now? I done all the comments.
@Tieske @outsinre

hanshuebner commented on lib/resty/rediscluster.lua:
```diff
@@ -76,6 +76,25 @@ local function check_auth(self, redis_client)
+            return true, nil
+        else
+            return nil, err
+        end
```
I think this logic would be easier to follow if the `err` would not be reassigned, but instead the function would return at the earliest point in time.  This seems to be a security relevant function that needs to be easy to review, so a refactoring is in order.

tysoekong replied:
Not sure what I should change here - it's only a little messy because I wanted to retain the backwards compatibility, if the config table is passed in with only "auth" field set.

What should I do?

Tieske replied:
@hanshuebner though I agree with you, the logic is identical to the check just before it (existing code).

hanshuebner replied:
```
local count, err = redis_client:get_reused_times()
if err then
  return nil, err
end

if count ~= 0 then
  # why?
  return true
end

if type(self.config.username) == "string" then
  return redis_client:auth(self.config.username, self.config.password)
end

# if we don't have a username, we just use the password (?)
return redis_client:auth(self.config.password)
```
This, and a couple of comments, would go a long way.  While at it, clarify the other code path as well.

outsinre replied:
In this place, we should check if the `username` is *default* or not. If it is set to *default*, we should return error.

Tieske replied:
my take;
```lua
local count, err = redis_client:get_reused_times()
if not count then
  return nil, err
elseif count ~= 0 then
  return true -- reusing connection, no need to authenticate again
end

if type(self.config.username) == "string" then
  return redis_client:auth(self.config.username, self.config.password)
end

# if we don't have a username, we just use the password (?)
return redis_client:auth(self.config.password)
```

tysoekong replied:
`default` is what redis 6.0+ now uses to refer to the old style "password only" user - so `default` is automatically assumed when ONLY a password is sent, with no accompanying username. It's a little weird.

Usually I would say just accept if the administrator puts "default" into the username box? Why not?!

outsinre replied:
> `default` is what redis 6.0+ now uses to refer to the old style "password only" user - so `default` is automatically assumed when ONLY a password is sent, with no accompanying username. It's a little weird.
> 
> Usually I would say just accept if the administrator puts "default" into the username box? Why not?!

As long as a user provides a `username` and `password` combination, we should do sanity check against the value of `username`.

hanshuebner replied:
Is it intentional that the code falls through if `count ~= null` or if username and password authentication return no err?  Generally, things are easier to understand when `if` statements are not deeply nested.  If a condition exists that allows returning, return.

outsinre replied:
> @outsinre any docs on that 'default' value? (I'd like to understand)

Here is a reference link: https://docs.konghq.com/hub/kong-inc/rate-limiting-advanced/

Check the `config.redis.username` part.

tyler-ball commented:
3.2 Feature freeze is next week - what is the status of this PR and KAG-401?

tysoekong commented:
I can get this merged on Monday, Thijs hopefully help me release it, then I can update the library + plugin in Kong-ee?

outsinre commented:
It would be better to add tests.

Tieske commented:
I just added a commit that fixes @hanshuebner comments wrt readibility. There are no existing auth tests in the suite, so demanding them here seems inappropriate to me (although having them would indeed be better).

tysoekong commented:
I made another fix, tested all four scenarios:
1. Kong 3.1.1.1 where it sends the argument `conf.auth` only - auth works, gets refused if wrong password
2. Kong 3.x.x.x master branch, where I have updated kong.redis package to send both `conf.username` and conf.password` - sent password only, works
3. Kong 3.x.x.x master branch, where I have updated kong.redis package to send both `conf.username` and conf.password` - sent username AND password, works
4. Kong 3.x.x.x master branch, where I have updated kong.redis package to send both `conf.username` and conf.password` - sent username as `default` - an error is thrown

Tieske reviewed (approved):
lgtm
